### PR TITLE
Implementation of CosFace algorithm for semantic categorical embeddings

### LINKED ISSUE #29: Fix categorical losses
Categorical variables are embedded with an online vocabulary, and because of the constant process of learning new vocabulary entries, the cross entropy loss consistently increases during the initial epochs of pre-training. Heavily imbalanced data among categorical variables prevents us from applying a heuristic hot-fix such as exclusion of losses considered when computing validation loss for a early stopping callback, and few options are immediately present, intuitive, and clean to implement. 

A warm start for a learned vocabulary is helpful, but that is not presently available for all cases, and so we must identify a generalizable solution which does not degrade performance, and which does not rely on any assumptions. 

I hypothesize that a algorithmic approach in which the tokenizer embedding vectors are all of unit magnitude, and in which the loss is of cosine similarity, will help to standardize the loss experienced, while also indirectly encouraging the model to learn more semantically "similar" embeddings for categorical values which are less discernible (and therefore more closely related).

### PR DESCRIPTION
**Pull request by Reid-Taylor**
By implementing an extension of the CosFace algorithm (Wang et al 2018) and the MCML penalty as a learning objective, this update for the categorical data type is aimed to provide more semantically interpretable embeddings while also more consistently providing monotonic convergence patterns during pre-training. The issue in the existing approach is that the expansion of new vocabulary for the online vocabulary causes an issue in which loss values grow over time as new vocabulary is learned, leading to EarlyStopping. Proofs provided in the original papers appendix support the default args, but this update does introduce two new hyper parameters relevant for categorical data types. 

We should question whether we want to implement this change as well for the set datatype, or if Categorical will now diverge from Set in this way. @granthamtaylor, thoughts?

Instead of hard-parameterizing embedding table to unit norm hypersphere, we opt to init the embedding table to this value, and then to normalize the values at each forward such that we only consider the projection of the embedding values upon the hypersphere. However, by not enforcing this hypersphere fixation on train batch end, we allow the model to more accurately leverage Adam optimizer steps and we provide greater stability more generally. There is also a geometric edge case in which the randomly initialized projection upon the hypersphere is encouraged to move to the extreme opposite end of the unit circle. By allowing semantic clusters to migrate freely, since loss is only considered by the angular measures of clusters and not radial measures, I hypothesize that we more freely allow semantic meaning to emerge under even extreme cases.

### CODE REVIEW

**Comment by Reid-Taylor** [comment]
Let me know if you want updates to docs for this change

**Comment by Reid-Taylor** [comment]
@granthamtaylor pinging this PR for testing & approval

**Comment by granthamtaylor** [comment]
I have it running now ! Looking for any regressions. Hopefully to be merged in tomorrow. Code otherwise looks good. 

We can bring the abstractions to some helper functions / Torch Modules to integrate with Set as well

**Comment by Reid-Taylor** [comment]
PR updated with optimizations discussed--re-norming the embedding table on epoch end, not on each call. Rebased to master

**Comment by granthamtaylor** [comment]
Taking another swing at this with `proofs` to help validate. Sorry for the delays here.
